### ISSUE TITLE
appveyor tests broken

### ISSUE THREAD

**Issue by GitMensch** [LADSoft/OrangeC#618]
Looking at [the current build log](https://ci.appveyor.com/project/LADSoft/orangec/builds/40170376/job/6ls5xxi1bi8un8bt?fullLog=true) we see a bunch of errors like the following:

~~~
Error: Undefined External '_BitScanForward' in module builtinsMicrosoft.c
Error: Undefined External '_BitScanReverse' in module builtinsMicrosoft.c
Error: Undefined External '_rotl16' in module builtinsMicrosoft.c
Error: Undefined External '_rotl8' in module builtinsMicrosoft.c
Error: Undefined External '_rotr16' in module builtinsMicrosoft.c
Error: Undefined External '_rotr8' in module builtinsMicrosoft.c
Error: Undefined External 'main' in module c0xpe.nas
	2 Errors, 0 Warnings
	Errors encountered, not creating output file
	occ localstaticinit.o
Error: Undefined External 'main' in module c0xpe.nas
	2 Errors, 0 Warnings
	Errors encountered, not creating output file
Error: Input file 'shared_ptr.o' does not exist.
occ Version 6.0.50.222 Copyright (C) LADSoft 2006-2021
Error: Input file 'simplelambda.o' does not exist.
	occ simplexcept.o
occ Version 6.0.50.222 Copyright (C) LADSoft 2006-2021
Error: Undefined External 'main' in module c0xpe.nas
	2 Errors, 0 Warnings
	Errors encountered, not creating output file
Error: Undefined External 'main' in module c0xpe.nas
	2 Errors, 0 Warnings
	Errors encountered, not creating output file
	olink /c /! /T:CON32 /olame.exe c0xpe.o bitstream.o common.o dct64_i386.o decode_i386.o encoder.o fft.o get_audio.o id3tag.o interface.o lame.o lame1.o lametime.o layer1.o layer2.o layer3.o mpglib_interface.o newmdct.o parse.o pcm.o portableio.o psymodel.o quantize.o quantize_pvt.o reservoir.o set_get.o tabinit.o tables.o takehiro.o timestatus.o tools.o util.o vbrquantize.o VbrTag.o version.o vorbis_interface.o clwin.l climp.l
	occ tuple_basic.o
Error: Input file 'localstaticinit.o' does not exist.
occ Version 6.0.50.222 Copyright (C) LADSoft 2006-2021
Error: Input file 'simplexcept.o' does not exist.
Error: Undefined External 'main' in module c0xpe.nas
	2 Errors, 0 Warnings
	Errors encountered, not creating output file
	occ strerror.o
~~~

Questions:

1. Why are those builds "green"?
2. (Really should only be tackled after the previous one!) Is there an issue with the setup of the tests or is there an actual error  leading to all those messages?
3. The tests should run with `occ /!` to reduce the size of the log, shouldn't it? I think it once did.

**Comment by LADSoft** [comment]
i thought I had fixed the green builds thing but apparently not....   I've pushed a change that simply attempts to make them go red and we will see if it fails.

@chuggafan the tests are failing because some of the GCC builtins aren't working the way they did before (such as __builtin_ctz and so forth).   Is there any chance your recent change to use an unordered map for the builtins could be affecting this?

I'm not sure we ever used /! on the tests, I know we did on the main compiles though.   I will change it once we get more of a handle as to what is going on here...

**Comment by GitMensch** [comment]
Related: 837fbe58d7e5f89de99878f4191f3893cbfd913c

**Comment by chuggafan** [comment]
> @chuggafan the tests are failing because some of the GCC builtins aren't working the way they did before (such as __builtin_ctz and so forth). Is there any chance your recent change to use an unordered map for the builtins could be affecting this?

It shouldn't.... *SIGH*, I'll test it when I get the chance today....

Theoretically they should all be located in an unordered_map and have lookups done against them so nothing wrong should happen.... I assumed green build = good too I guess.

*SIGH*

**Comment by LADSoft** [comment]
yeah sorry about that, I really thought I had fixed the green build problem.

**Comment by chuggafan** [comment]
OK: so it's likely the code changes I did (likely) broke something else, probably has to do with the hash system I set up not actually functioning despite literally reading it off the page on wikipedia and the public domain'd source code... will look into comparing the two outputs and seeing the difference.

I do however find it interesting that an artifact(?) of the builtins actually gets through, specifically al gets promoted to boolean.

I will likely just test this by comparing the name of the AMode going through and what actually happens, since the ONLY changes that were made (other than to stdint) in that commit were to the builtins file.

*YAY*

**Comment by chuggafan** [comment]
Ok so I've confirmed it's the map code, the question is: what? 7108444667618756687 is pretty obviously different to 1712599851734136393 but what causes one vs the other to be the way it is?

I must investigate deeper.

**Comment by chuggafan** [comment]
Ok, so either builtin_map is bad or something else is going on: 7108444667618756687 is the hash representation of _rotr, and the builtins map does hash it to this, as tested by the following: 

```cpp
    printf("Name searching for builtin: %s\n", name);
    if (!strcmp(name, "_rotr"))
    {
        printf("Internal representation from just fnv1a64: %lld\n", builtin_map.hash_function()(name));
        printf("Internal representation hardcoded: %lld\n", builtin_map.hash_function()("_rotr"));
        for (auto& thing : builtin_map)
        {
            printf("Character representation of all map internals: %s\n", thing.first);
            printf("Hash representation of the same: %lld", builtin_map.hash_function()(thing.first));
        }
    }
```

Everything here demonstrates that it *SHOULD WORK*, which means I think there's an underlying issue either with MSVC or libcxx causing this to fail spectacularly.  However, the find and then find != end fails, meaning *something is wrong*.

I dunno from this point on, all I know is that all signs point to life where there should be but there ain't.

**Comment by LADSoft** [comment]
i can look at it tonite if you'd like?   But I'm glad you at least confirmed this as the issue, gives us something concrete to look at.   I honestly wasn't sure I hadn't broken something entirely unrelated lol.

**Comment by LADSoft** [comment]
I looked a liitle, maybe you need an operator== or something similar?   apparently needed for collisions, although maybe also the hash table needs to be allocated at a larger size to avoid that.    Anyway try reading this:

https://stackoverflow.com/questions/17016175/c-unordered-map-using-a-custom-class-type-as-the-key

**Comment by chuggafan** [comment]
Yep, that was the issue, it checked for hash, which worked, and then == the addresses, which failed.

I fixed it by doing:

```cpp
class str_eql
{
    public:
    bool operator()(const char* a, const char* b) const
    {
        return !strcmp(a, b);
    }
};
```
and:

`std::unordered_map<const char*, BUILTIN, fnv1a64, str_eql>`

Will just submit a PR for this as it's a simpler fix than I expected.

**Comment by LADSoft** [comment]
cool!   last I checked with my latest fix to the build files the builds seemed to be successfully failing, so we should be able to see this work when you submit :smile:

**Comment by LADSoft** [comment]
the build seems to be doing the builtins ok now, from what I could tell looking at the log the current hangup is a missing SIZE_MAX and UINT32_MAX.

**Comment by LADSoft** [comment]
@chuggafan I wasn't sure what to do about this.   I know you'd read the standard ecently and knew how it should work, and probably this issue is related to another of your recent changes.    But I'm not sure what change I should make to keep it within the standard.  Should the header change?   Or should I change the tests to match the header?

**Comment by LADSoft** [comment]
@chuggafan,   this may have only been a problem for the PR you made, it seems to compile fine on my computer.   I've got another appveyor build going now...

**Comment by LADSoft** [comment]
well it works on my computer, but doesn't work on appveyor.   And moreover it is failing on C language files, not C++ language files.   I'll try to figure out what is going on this afternoon.

**Comment by GitMensch** [comment]
We now have a failure in the CI when the tests fail, which is good. We also have smaller logs because of using the `/!` option.

Now the "only thing" missing is getting the builds working again.
The following ones look very suspicious "could not open ... for input":

~~~
Error (0): Could not open lookahead.c for input
	occ /c /9 /! -omacroblock.o macroblock.c
Error (0): Could not open macroblock.c for input
	occ /c /9 /! -ointernal.o internal.c
Error (0): Could not open internal.c for input
	occ /c /9 /! -ogetopt.o getopt.c
	occ /c /9 /! -omatroska.o matroska.c
Error (0): Could not open matroska.c for input
Error (0): Could not open getopt.c for input
	occ /c /9 /! -omatroska_ebml.o matroska_ebml.c
	occ /c /9 /! -omc.o mc.c
Error (0): Could not open matroska_ebml.c for input
Error (0): Could not open mc.c for input
	occ /c /9 /! -ome.o me.c
	occ /c /9 /! -omvpred.o mvpred.c
Error (0): Could not open me.c for input
	occ /c /! -opsymodel.o psymodel.c
Error (0): Could not open mvpred.c for input
	occ /c /! -oquantize.o quantize.c
	occ /c /! -oquantize_pvt.o quantize_pvt.c
	occ -D__SIZE_TYPE__=unsigned /1 /c /! stdatomic-lockfree.c
Error (0): Could not open stdatomic-lockfree.c for input
	occ -D__SIZE_TYPE__=unsigned /1 /c /! stdatomic-op-1.c
	occ -D__SIZE_TYPE__=unsigned /1 /c /! stdatomic-op-2.c
Error (0): Could not open stdatomic-op-2.c for input
	occ /c /! stlcontainers.cpp
Error (0): Could not open stdatomic-op-1.c for input
	occ /c /! stlcontainers2.cpp
	occ /c /! strerror.cpp
	occ /c /! tuple_basic.cpp
	occ /c /! simplexcept.cpp
	occ /c /9 /! -ooraw.o oraw.c
Error (0): Could not open oraw.c for input
	occ /c /9 /! -oosdep.o osdep.c
Error (0): Could not open osdep.c for input
	occ /c /9 /! -opixel.o pixel.c
Error (0): Could not open pixel.c for input
	occ /c /! -oset_get.o set_get.c
	occ /c /! -oreservoir.o reservoir.c
	occ /c /! -otabinit.o tabinit.c
	occ /c /! -otables.o tables.c
	occ /c /! -otakehiro.o takehiro.c
omake: aborting due to timeout
Error makefile(21): commands returned error code 2
	occ -D__SIZE_TYPE__=unsigned /1 /c /! stdatomic-op-4.c
Error (0): Could not open stdatomic-op-4.c for input
	occ -D__SIZE_TYPE__=unsigned /1 /c /! stdatomic-store-1.c
Error (0): Could not open stdatomic-store-1.c for input
	occ -D__SIZE_TYPE__=unsigned /1 /c /! stdatomic-op-5.c
Error (0): Could not open stdatomic-op-5.c for input
	occ -D__SIZE_TYPE__=unsigned /1 /c /! stdatomic-op-3.c
Error (0): Could not open stdatomic-op-3.c for input
~~~

**Comment by LADSoft** [comment]
yeah those are actually artifacts of a previous problem.   The actual problem is that SIZE_MAX and UINT32_MAX aren't defined when compiling certain C language files.   I'll chase it down this afternoon.   

For bonus points I probably *ought* to figure out why a failure cascades like this lol...   I suspect it has something to do with failing while doing parallel processing...

**Comment by GitMensch** [comment]
> I suspect it has something to do with failing while doing parallel processing...

Then this should be tackled first as an omake issue - as soon as any target failed the one that started it (either as a dependency or direct invocation) must directly fail (leaving `omake -k` aside) and no more parallel targets should be started as soon as one target finished with an error.

**Comment by chuggafan** [comment]
> @chuggafan I wasn't sure what to do about this. I know you'd read the standard ecently and knew how it should work, and probably this issue is related to another of your recent changes. But I'm not sure what change I should make to keep it within the standard. Should the header change? Or should I change the tests to match the header?

Do you know which header the two are supposed to live in but aren't? There's actually a clib test group (test, std, depr, depr.c.headers) for the libcxx and we can test against it to make sure our headers provide expected functions such as `int_least*_t` and others... I think testing against this and seeing what files fail and adding what doesn't exist there would be good because we're testing for it *ANYWAYS*.

Anywho: The actual fix is, since I finally got around to checking, removing the __STDC_LIMIT_MACROS and __STDC_CONSTANT_MACROS macros from stdint.h and just always defining the two groups.

**Comment by LADSoft** [comment]
@chuggafan, if the header tests are in with the rest of the libcxx tests I guess I'll stumble on them before too long.   Gotta get past my constexpr woes first though.    Meanwhile, I did get stdint.h to work based on your suggestion.

@GitMensch, I think it might compile ok now.   I did improve the situation so omake will bail more quickly now.   It isn't perfect though...   guess we'll see how I did in an hour or so.
